fjpgtt/pollitos-school
# Martin-school: Add REST services for management of POLLITOS SCHOOL

Change Summary:

SchoolControlApplication organized by packages and files, taking into account CRUD endpoints for grades and typical elements in a Spring Boot application with entities such as School, Student, Course, Grade, etc.

1. SchoolControlApplication.java: It is the main class that starts the Spring Boot application. Contains the main method that runs the application.
2. controller/: Contains the REST controllers that expose the APIs. In your case, the controllers for Grade, Student, Course, and School.
3. service/: Contains the business logic of the entities, such as CRUD operations and data validation.
4. model/: Contains the entities or data models (for example, Grade, Student, Course), which are mapped to database tables with JPA.
5. config/: Contains global configurations of the beans application

This PR introduces new REST services for managing grades, students, courses in the SchoolControlApplication system, facilitating the creation, reading, updating and deletion of elements associated with students within 2 schools GerardoInstitute & ZetCollege. These services allow you to manage grades, courses and students belonging to a particular school. 

Loading default data with the 'populated' profile: Updated the project configuration to ensure that when you run the populated profile, preconfigured test data is loaded into the Grade, Student, and School entities, with example initial values, which which facilitates the testing and integration process in development environments.

Latest changes

"This shouldn't be a bean, because a bean is a reusable component."
* Remove @Component Student, School, Grade, Course

"For POST and PUT request you should use @RequestBody, because in @RequestParam you have to put the info in the URL and this is dangerous"
    *studentGrades: Since this method only receives a @PathVariable and a @RequestParam, I do not consider @RequestBody necessary, since a JSON body is not being sent.
    *Some methods were updated to @RequestBody and the previous ones were preserved to exemplify another way of doing it by @RequestParam  *instructions were added to be able to test the added points
          PUT update a student
          POST to create a new student
          PUT update a course
          POST create a course
          POST create grade

The following endpoints were implemented:
Create student
Create course
Create a grade
Get all the students
Get all the courses
Get all the grades from a specific student
Edit the course information
Edit the student information
Delete all the grades of a student
Delete all the grades of a course

Extra Endpoints con  @RequestBody 
PUT update a student
POST to create a new student
PUT update a course
POST create a course
POST create grade

Extra Endpoint 
Get all the grades

How to do tests (POSTMAN): https://www.postman.com/downloads/

- Ensure that the GET, POST, PUT and DELETE methods respond as expected.
- Validate that the data uploaded when running the POPULATED profile includes sample grades for students, courses, and schools.
- Perform tests with preloaded data for ease

Create student
POST
http://localhost:8080/school/GerardoInstitute/students?id=7&firstName=Juan&lastName=Perez&age=66
http://localhost:8080/school/ZetCollege/students?id=7&firstName=Juan&lastName=Perez&age=66

Create course
POST
http://localhost:8080/school/GerardoInstitute/courses?id=1&name=Roberto&professorName=Perez
http://localhost:8080/school/ZetCollege/courses?id=1&name=Roberto&professorName=Perez

Create a grade
POST
http://localhost:8080/school/GerardoInstitute/grade?score=90&studentId=1&courseId=2
http://localhost:8080/school/ZetCollege/grade?score=90&studentId=1&courseId=2

Get all the students
http://localhost:8080/school/GerardoInstitute/getStudents
http://localhost:8080/school/ZetCollege/getStudents

Get all the courses
http://localhost:8080/school/GerardoInstitute/getCourses
http://localhost:8080/school/ZetCollege/getCourses

Get all the grades from a specific student
http://localhost:8080/school/GerardoInstitute/getGrades
http://localhost:8080/school/ZetCollege/getGrades

Edit the course information
PUT
http://localhost:8080/school/GerardoInstitute/course?id=1&name=Roberto88&professorName=Perez88
http://localhost:8080/school/ZetCollege/course?id=1&name=Roberto88&professorName=Perez88

Edit the student information
PUT
http://localhost:8080/school/GerardoInstitute/students?id=99&firstName=Juan99&lastName=Perez99&age=6699
http://localhost:8080/school/ZetCollege/students?id=99&firstName=Juan99&lastName=Perez99&age=6699

Delete all the grades of a student
http://localhost:8080/school/GerardoInstitute/studentGradesDelete?id=1
http://localhost:8080/school/ZetCollege/studentGradesDelete?id=3

Delete all the grades of a course
http://localhost:8080/school/GerardoInstitute/courseGradesDelete?id=1
http://localhost:8080/school/ZetCollege/courseGradesDelete?id=3

Extra Endpoints con  @RequestBody 

PUT update a student
http://localhost:8080/GerardoInstitute/bodyStudent
http://localhost:8080/ZetCollege/bodyStudent
{
    "id": 1,
    "firstName": "Juan",
    "lastName": "Pérez",
    "age": 19
}


POST to create a new student
http://localhost:8080/school/GerardoInstitute/student
http://localhost:8080/school/ZetCollege/student
{
    "id": 5,
    "firstName": "María",
    "lastName": "García",
    "age": 22
}

PUT update a course
http://localhost:8080/school/GerardoInstitute/getCourses
http://localhost:8080/school/ZetCollege/getCourses
{
    "id": 1,
    "name": "Historia",
    "professorName": "Prof. Sánchez"
}

POST create a course
http://localhost:8080/school/GerardoInstitute/bodyCourse
http://localhost:8080/school/ZetCollege/bodyCourse
{
    "id": 99,
    "name": "Historia",
    "professorName": "Prof. Sánchez"
}

POST create grade
http://localhost:8080/school/GerardoInstitute/bodyGrade
http://localhost:8080/school/ZetCollege/bodyGrade
{
    "score": 95,
    "studentId": 4,
    "courseId": 2
}

Extra Endpoint 
Get all the grades
http://localhost:8080/school/GerardoInstitute/getGrades
http://localhost:8080/school/ZetCollege/getGrades

Criteria to evaluate:

Creation of beans ✔️ 
Usage of the annotation in the Controller layer ✔️ 
Usage of Spring and Spring boot annotation ✔️ 
Implementation of logs ✔️ 
Naming convention ✔️ 

Extra points:
Creation of unit test or integration ⏳ 
Correct usage of interfaces and abstract classes ⏳ 
Proper instructions in the PR (description, title, steps to test) ✔️
